src/tester.rs

```diff
@@ -0,0 +1,446 @@
+pub trait Testable<S>: 'static
+where
+    S: Sample,
+{
+    fn result(&self, v: S::Output) -> TestResult;
+
+    fn test_once(&self, s: &S, rng: &mut Random) -> TestResult
+    where
+        S::Output: Clone + Debug,
+    {
+        let v = Sample::generate(s, rng);
+        let r = self.result(v.clone());
+        match r.status {
+            Pass | Discard => r,
+            Fail => {
+                error!("{:?}", r);
+                self.shrink(s, r, v)
+            }
+        }
+    }
+
+    fn shrink(&self, s: &S, r: TestResult, v: S::Output) -> TestResult
+    where
+        S::Output: Clone + Debug,
+    {
+        trace!("shrinking {:?}", v);
+        let mut result = r;
+        let mut it = s.shrink(v);
+        let iterations = 10_000_000;
+
+        for _ in 0..iterations {
+            let sv = it.next();
+            if let Some(sv) = sv {
+                let r_new = self.result(sv.clone());
+                if r_new.is_failure() {
+                    trace!("shrinking {:?}", sv);
+                    result = r_new;
+                    it = s.shrink(sv);
+                }
+            } else {
+                return result;
+            }
+        }
+
+        trace!(
+            "halting shrinkage after {} iterations with: {:?}",
+            iterations,
+            result
+        );
+
+        result
+    }
+}
+
+impl From<bool> for TestResult {
+    fn from(value: bool) -> TestResult {
+        TestResult::from_bool(value)
+    }
+}
+
+impl From<()> for TestResult {
+    fn from(_: ()) -> TestResult {
+        TestResult::passed()
+    }
+}
+
+impl<A, E> From<Result<A, E>> for TestResult
+where
+    TestResult: From<A>,
+    E: Debug + 'static,
+{
+    fn from(value: Result<A, E>) -> TestResult {
+        match value {
+            Ok(r) => r.into(),
+            Err(err) => TestResult::error(format!("{:?}", err)),
+        }
+    }
+}
+
+macro_rules! testable_fn {
+    ($($name: ident),*) => {
+
+impl<T: 'static, S, $($name),*> Testable<S> for fn($($name),*) -> T
+where
+    TestResult: From<T>,
+    S: Sample<Output=($($name),*,)>,
+    ($($name),*,): Clone,
+    $($name: Debug + 'static),*
+{
+    #[allow(non_snake_case)]
+    fn result(&self, v: S::Output) -> TestResult {
+        let ( $($name,)* ) = v.clone();
+        let f: fn($($name),*) -> T = *self;
+        let mut r = <TestResult as From<Result<T, String>>>::from(safe(move || {f($($name),*)}));
+
+        {
+            let ( $(ref $name,)* ) = v;
+            r.arguments = format!("{:?}", &($($name),*));
+        }
+        r
+    }
+}}}
+
+testable_fn!(A);
+testable_fn!(A, B);
+testable_fn!(A, B, C);
+testable_fn!(A, B, C, D);
+testable_fn!(A, B, C, D, E);
+testable_fn!(A, B, C, D, E, F);
+testable_fn!(A, B, C, D, E, F, G);
+testable_fn!(A, B, C, D, E, F, G, H);
+
+fn safe<T, F>(fun: F) -> Result<T, String>
+where
+    F: FnOnce() -> T,
+    F: 'static,
+    T: 'static,
+{
+    panic::catch_unwind(panic::AssertUnwindSafe(fun)).map_err(|any_err| {
```

Comment:
ooh, I've never played with `std::panic` stuff, fun!

Reply:
Haha yeah. This was stolen from `quickcheck`. Useful for when a test run panics (e.g. an assert fails); we don't want to bubble the panic up, we want to treat it as a failure during the shrinking procedure.